maskrcnn_benchmark/modeling/poolers.py

```diff
@@ -116,6 +116,6 @@ def forward(self, x, boxes):
         for level, (per_level_feature, pooler) in enumerate(zip(x, self.poolers)):
             idx_in_level = torch.nonzero(levels == level).squeeze(1)
             rois_per_level = rois[idx_in_level]
-            result[idx_in_level] = pooler(per_level_feature, rois_per_level)
+            result[idx_in_level] = pooler(per_level_feature, rois_per_level).to(dtype)
```

Comment:
doesn't the `amp.float_function` wrap back the values to fp16 after they are computed?

Reply:
I'd need to run again to work out exactly what case was happening here, but the type change was not happening correctly and I had to manually cast here to prevent errors

Reply:
Quick question: is this casting still relevant?

Reply:
Yes: `result[idx_in_level]` is expected to be in fp16 (as it's the same precision as the input), but the pooler returns fp32 (explicitly, as the code hasn't had fp16 support added). To get around this the result from the pooler needs to be casted

Reply:
But I thought that `amp.float_function` would:
1 - cast to float
2 - compute
3 - cast back to fp16

Or is my understanding of it wrong?

Reply:
By my understanding this is an artifact of how MRCNN is constructed. The input to Pooler is cast to `half`. It initialises the output for pooling to that type. Then the Pooler module calls the `ROIPool` module, which casts _it's_ input to `float`. It doesn't explicitly cast back to `half` (the next module in line would do this for you), so it tries to assign a `float` tensor to a `half` result. This only occurs because the poolers are members of a Module, rather than composed with that module using `nn.Sequential` or otherwise

Reply:
Ok.
I think the solution I'd potentially do myself (while support for fp16 is not present in the core pooler functions) is to just cast in the C++ side to float and cast back if the type is fp16.
But I suppose this is not really a hard-requirement here (but would make things cleaner)

Reply:
Would you like a comment added explaining the current need for the cast? (along with a TODO for the full (fp16 support in pooling) if you so desire)

Reply:
Because this is a strange case -- you're in fp16-land, allocate your output as fp16, then run something that has to cast up to fp32. There's no cast back automatically (you're calling from a module, so there's no module boundary to cause a cast) so it has to be done manually. If lines 111-115 didn't exist, and you inferred the type from the return type of the pooling call this explicit wouldn't have to be there.

Reply:
I understand it's a little weird, but that's where the code is right now -- if you want an explicit (non-AMP) fp16 version that can also be done, but it'll be more invasive and can't be done before the new year after I get back